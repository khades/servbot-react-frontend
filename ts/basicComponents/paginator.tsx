import classnames from "classnames";
import * as React from "react";

export interface IPaginatorState {
    lastPage: boolean;
    firstPage: boolean;
}

export interface IPaginatorProps {
    pages: number;
    page: number;
    setPage: (page: number) => void;
}

export default class Paginator extends React.Component<IPaginatorProps, IPaginatorState> {
    public static getDerivedStateFromProps(props: IPaginatorProps, state: IPaginatorState) {
        return {
            firstPage: props.page === 1,
            lastPage: props.page === props.pages,
        };
    }

    public static generateButtonsRange(pages: number, page: number) {
        const paginatorRange: number = 2;
        const messedPages = [
            paginatorRange - 2,
            paginatorRange - 1,
            paginatorRange,
            paginatorRange + 1,
            paginatorRange + 2];
        return messedPages
            .map((f) => page - paginatorRange + f)
            .filter((f) => f > 0 && f < pages + 1);
    }

    constructor(props: IPaginatorProps) {
        super(props);

        this.state = {
            firstPage: props.page === 1,
            lastPage: props.page === props.pages,

        };
    }

    public render() {
        return (
            <div className="paginator">
                {this.generatePrevButtons()}
                {this.generateButtons()}
                {this.generateNextButtons()}
            </div>
        );
    }

    private generateButtons = () => {
        const pages = Paginator.generateButtonsRange(this.props.pages, this.props.page);
        return (
            <React.Fragment>
                {pages.map((page) => this.generateButton(page))}
            </React.Fragment>
        );
    }

    private generateButton = (value: number) => {
        const classes = classnames({
            "paginator__item": true,
            "paginator__item--desktop": true,
            "paginator__item--selected": this.props.page === value,
        });
        const onclick = () => {
            this.props.setPage(value);
        };
        return (
            <div className={classes} data-value={value} key={value.toString()} onClick={onclick}>
                <div className="paginator__item-content">
                    {value.toString()}
                </div>
            </div>
        );
    }
    private generatePrevButtons() {
        const firstButtonClasses = classnames({
            "paginator__item": true,
            "paginator__item--desktop": true,
            "paginator__item--selected": this.state.firstPage,
        });

        const prevButtonClasses = classnames({
            "paginator__item": true,
            "paginator__item--inactive": this.state.firstPage,
        });

        return (
            <React.Fragment>
                <div className={firstButtonClasses} data-value={"first"} key="first" onClick={this.goToFirst}>
                    <div className="paginator__item-content">
                        "❮❮"
                    </div>
                </div>
                <div
                    className={prevButtonClasses}
                    data-value={"prev"}
                    key="prev"
                    onClick={this.goToPrevious}
                >
                    <div className="paginator__item-content">
                        "❮"
                    </div>
                </div>
            </React.Fragment>
        );
    }
    private generateNextButtons = () => {
        const lastButtonClasses = classnames({
            "paginator__item": true,
            "paginator__item--desktop": true,
            "paginator__item--selected": this.state.lastPage,
        });

        const nextButtonClasses = classnames({
            "paginator__item": true,
            "paginator__item--inactive": this.state.lastPage,
        });

        return (
            <React.Fragment>
                <div
                    className={nextButtonClasses}
                    data-value={"next"}
                    key="next"
                    onClick={this.goToNext}
                >
                    <div className="paginator__item-content">
                        "❯"
                    </div>
                </div>
                <div
                    className={lastButtonClasses}
                    data-value={"last"}
                    key="last"
                    onClick={this.goToLast}
                >
                    <div className="paginator__item-content">
                        "❯❯"
                    </div>
                </div>

            </React.Fragment >
        );
    }

    private goToNext = () => {
        if (this.state.lastPage !== true) {
            this.props.setPage(this.props.page + 1);
        }
    }

    private goToLast = () => {
        if (this.state.lastPage !== true) {
            this.props.setPage(this.props.pages);
        }
    }

    private goToFirst = () => {
        if (this.state.firstPage !== true) {
            this.props.setPage(1);
        }
    }

    private goToPrevious = () => {
        if (this.state.firstPage !== true) {
            this.props.setPage(this.props.page - 1);
        }
    }
}
