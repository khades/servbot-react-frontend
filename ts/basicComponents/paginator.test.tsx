import * as React from "react";
import { create } from "react-test-renderer";
import { IPaginatorProps } from "./paginator";
import Paginator from "./paginator";
describe("Paginator", () => {
    it("should properly calculate ranges", () => {
        expect(Paginator.generateButtonsRange(14, 3)).toEqual([1, 2, 3, 4, 5]);
        expect(Paginator.generateButtonsRange(14, 1)).toEqual([1, 2, 3]);
        expect(Paginator.generateButtonsRange(14, 2)).toEqual([1, 2, 3, 4]);
        expect(Paginator.generateButtonsRange(14, 6)).toEqual([4, 5, 6, 7, 8]);
        expect(Paginator.generateButtonsRange(14, 11)).toEqual([9, 10, 11, 12, 13]);
        expect(Paginator.generateButtonsRange(14, 12)).toEqual([10, 11, 12, 13, 14]);
        expect(Paginator.generateButtonsRange(14, 13)).toEqual([11, 12, 13, 14]);
        expect(Paginator.generateButtonsRange(14, 14)).toEqual([12, 13, 14]);
        expect(Paginator.generateButtonsRange(4, 1)).toEqual([1, 2, 3]);
        expect(Paginator.generateButtonsRange(4, 2)).toEqual([1, 2, 3, 4]);
        expect(Paginator.generateButtonsRange(4, 3)).toEqual([1, 2, 3, 4]);
        expect(Paginator.generateButtonsRange(4, 4)).toEqual([2, 3, 4]);
        expect(Paginator.generateButtonsRange(3, 1)).toEqual([1, 2, 3]);
        expect(Paginator.generateButtonsRange(3, 2)).toEqual([1, 2, 3]);
        expect(Paginator.generateButtonsRange(3, 3)).toEqual([1, 2, 3]);
        expect(Paginator.generateButtonsRange(2, 1)).toEqual([1, 2]);
        expect(Paginator.generateButtonsRange(2, 2)).toEqual([1, 2]);
        expect(Paginator.generateButtonsRange(1, 1)).toEqual([1]);

    });
    it("should properly render for page 1 and pages 14", () => {
        const props: IPaginatorProps = {
            getPage: () => 1,
            getPages: () => 14,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });
    it("should properly render for page 3 and pages 14", () => {
        const props: IPaginatorProps = {
            getPage: () => 3,
            getPages: () => 14,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
    });
    it("should properly render for page 6 and pages 14", () => {
        const props: IPaginatorProps = {
            getPage: () => 6,
            getPages: () => 14,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });
    it("should properly render for page 11 and pages 14", () => {
        const props: IPaginatorProps = {
            getPage: () => 11,
            getPages: () => 14,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });

    it("should properly render for page 14 and pages 14", () => {
        const props: IPaginatorProps = {
            getPage: () => 14,
            getPages: () => 14,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });

    it("should properly render for page 1 and pages 4", () => {
        const props: IPaginatorProps = {
            getPage: () => 1,
            getPages: () => 4,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });
    it("should properly render for page 3 and pages 4", () => {
        const props: IPaginatorProps = {
            getPage: () => 1,
            getPages: () => 4,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });

    it("should properly render for page 1 and pages 2", () => {
        const props: IPaginatorProps = {
            getPage: () => 1,
            getPages: () => 2,
            setPage: (input: number) => input,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).toJSON();
        expect(paginator).toMatchSnapshot();
    });
    it("should properly react to numeric button presses", () => {
        let page = 11;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === 13);
        button.props.onClick();
        expect(page).toBe(13);

    });
    it("should properly react to arrow button presses - \"<\" case", () => {
        let page = 11;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "prev");
        button.props.onClick();
        expect(page).toBe(10);
    });

    it("should properly react to arrow button presses - \"<\" case, already first item", () => {
        let page = 1;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "prev");
        button.props.onClick();
        expect(page).toBe(1);
    });

    it("should properly react to arrow button presses - \"<<\" case", () => {
        let page = 11;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "first");
        button.props.onClick();
        expect(page).toBe(1);
    });

    it("should properly react to arrow button presses - \"<<\" case, already first item", () => {
        let page = 1;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "first");
        button.props.onClick();
        expect(page).toBe(1);
    });

    it("should properly react to arrow button presses - \">\" case", () => {
        let page = 11;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "next");
        button.props.onClick();
        expect(page).toBe(12);
    });

    it("should properly react to arrow button presses - \">\" case, already last item", () => {
        let page = 14;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "next");
        button.props.onClick();
        expect(page).toBe(14);
    });

    it("should properly react to arrow button presses - \">>\" case", () => {
        let page = 11;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "last");
        button.props.onClick();
        expect(page).toBe(14);
    });

    it("should properly react to arrow button presses - \">>\" case, already last item", () => {
        let page = 14;
        const props: IPaginatorProps = {
            getPage: () => page,
            getPages: () => 14,
            setPage: (value: number) => page = value,
        };

        const paginator = create(
            <Paginator {...props} />,
        ).root;
        const button = paginator.find((el) => el.props["data-value"] === "last");
        button.props.onClick();
        expect(page).toBe(14);
    });
});
