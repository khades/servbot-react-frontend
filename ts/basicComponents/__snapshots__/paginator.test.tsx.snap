// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator should properly render for page 1 and pages 2 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--inactive"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={1}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      1
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={2}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      2
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 1 and pages 4 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--inactive"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={1}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      1
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={2}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      2
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={3}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      3
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 1 and pages 14 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--inactive"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={1}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      1
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={2}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      2
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={3}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      3
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 3 and pages 4 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--inactive"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={1}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      1
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={2}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      2
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={3}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      3
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 6 and pages 14 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={4}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      4
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={5}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      5
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={6}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      6
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={7}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      7
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={8}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      8
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 11 and pages 14 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={9}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      9
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={10}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      10
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={11}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      11
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={12}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      12
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={13}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      13
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;

exports[`Paginator should properly render for page 14 and pages 14 1`] = `
<div
  className="paginator"
>
  <div
    className="paginator__item paginator__item--desktop"
    data-value="first"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮❮"
    </div>
  </div>
  <div
    className="paginator__item"
    data-value="prev"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❮"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={12}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      12
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop"
    data-value={13}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      13
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value={14}
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      14
    </div>
  </div>
  <div
    className="paginator__item paginator__item--inactive"
    data-value="next"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯"
    </div>
  </div>
  <div
    className="paginator__item paginator__item--desktop paginator__item--selected"
    data-value="last"
    onClick={[Function]}
  >
    <div
      className="paginator__item-content"
    >
      "❯❯"
    </div>
  </div>
</div>
`;
